conduit/metrics.py

```diff
@@ -0,0 +1,249 @@
+from enum import Enum
+from functools import partial, wraps
+from typing import List, Optional, Protocol, Tuple, TypeVar, Union
+
+import torch
+from torch import Tensor
+from typing_extensions import runtime_checkable
+
+__all__ = [
+    "Comparator",
+    "accuracy",
+    "accuracy_per_subclass",
+    "groupwise_metric",
+    "hard_prediction",
+    "max_difference_1d",
+    "precision_at_k",
+    "robust_accuracy",
+    "robust_gap",
+    "subclass_balanced_accuracy",
+    "subclasswise_metric",
+    "tnr_per_subclass",
+    "tpr_differences",
+    "tpr_per_subclass",
+]
+
+
+@torch.no_grad()
+def hard_prediction(logits: Tensor) -> Tensor:
+    logits = torch.atleast_2d(logits.squeeze())
+    return (logits > 0).long() if logits.ndim == 1 else logits.argmax(dim=1)
+
+
+@torch.no_grad()
+def accuracy(y_pred: Tensor, *, y_true: Tensor) -> Tensor:
+    y_pred = torch.atleast_2d(y_pred.squeeze())
+    if len(y_pred) != len(y_true):
+        raise ValueError("'logits' and 'targets' must match in size at dimension 0.")
+    preds = hard_prediction(y_pred)
+    return (preds == y_true).float().mean()
+
+
+@torch.no_grad()
+def precision_at_k(
+    y_pred: Tensor, *, y_true: Tensor, top_k: Union[int, Tuple[int, ...]] = (1,)
+) -> List[Tensor]:
+    """Computes the accuracy over the k top predictions for the specified values of k"""
+    if isinstance(top_k, int):
+        top_k = (top_k,)
+    maxk = max(top_k)
+    batch_size = y_true.size(0)
+    _, pred = y_pred.topk(k=maxk, dim=1, largest=True, sorted=True)
+    pred = pred.t()
+    correct = pred.eq(y_true.view(1, -1).expand_as(pred))
+
+    res: List[Tensor] = []
+    for k in top_k:
+        correct_k = correct[:k].contiguous().view(-1).float().sum(0, keepdim=True)
+        res.append(correct_k.mul_(100.0 / batch_size))
+    return res
+
+
+R = TypeVar("R", Tensor, Tuple[Tensor, Union[Tensor, slice]], covariant=True)
+
+
+@runtime_checkable
```

Comment:
There is little point in making this runtime checkable because it will just check whether the object is callable.